src/engine/strat_engine/engine.rs

```diff
@@ -124,6 +140,58 @@ impl Engine for StratEngine {
         Ok(uuid)
     }
 
+    /// Evaluate a device node & devicemapper::Device to see if it's a valid
+    /// stratis device.  If all the devices are present in the pool and the pool isn't already
+    /// up and running, it will get setup and the pool uuid will be returned.
+    ///
+    /// Returns an error if the status of the block device can not be evaluated.
+    fn block_evaluate(&mut self,
+                      dev_node: PathBuf,
+                      device: Device)
+                      -> EngineResult<Option<PoolUuid>> {
+        if let Some(pool_uuid) = is_stratis_device(&dev_node)? {
+            if self.pools.contains_uuid(pool_uuid) {
+                warn!("Pool {} is already known, ignoring!", pool_uuid);
+                Ok(None)
+            } else {
+                self.incomplete_pools
+                    .entry(pool_uuid)
+                    .or_insert_with(HashMap::new)
+                    .insert(device, dev_node);
+
+                let devices = self.incomplete_pools
+                    .get(&pool_uuid)
+                    .expect("We previously inserted pool uuid!")
+                    .clone();
```

Comment:
```
error[E0502]: cannot borrow `self.incomplete_pools` as mutable because it is also borrowed as immutable
   --> src/engine/strat_engine/engine.rs:171:25
    |
162 |                 let devices = self.incomplete_pools
    |                               --------------------- immutable borrow occurs here
...
171 |                         self.incomplete_pools.remove(&pool_uuid);
    |                         ^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
...
188 |             }
    |             - immutable borrow ends here
```

Reply:
Oh, I see. You might be better off removing the devices here, and reinserting them below if necessary. Beats a deep copy.

Reply:
My previous implementation didn't require a clone...